File: benchmarks/micro/ExplicitArithmeticRangePruning.c
#include <stdint.h>
#include <unistd.h>

static int32_t saturating_add_i32(int32_t left, int32_t right) {
    int64_t result = (int64_t)left + (int64_t)right;
    if (result > INT32_MAX) {
        return INT32_MAX;
    }

    if (result < INT32_MIN) {
        return INT32_MIN;
    }

    return (int32_t)result;
}

static int64_t score(int32_t left, int32_t right, int64_t salt) {
    int32_t saturated = saturating_add_i32(left, right);
    int32_t wrapped = left + right;

    if (saturated > 15) {
        return salt - 10000;
    }

    if (wrapped > 15) {
        return salt - 20000;
    }

    return (((int64_t)saturated * 37) + ((int64_t)wrapped * 17) + salt) % 1000003;
}

int main(void) {
    int64_t total = (int64_t)getpid();

    for (int32_t i = 0; i < 200000; i += 1) {
        total += score(i % 11, i % 6, total % 97);
        total %= 1000000007;
    }

    return total <= 0 ? 1 : 0;
}

File: benchmarks/micro/ExplicitArithmeticRangePruning.rs
fn score(left: i32, right: i32, salt: i64) -> i64 {
    let saturated = left.saturating_add(right);
    let wrapped = left.wrapping_add(right);

    if saturated > 15 {
        return salt - 10_000;
    }

    if wrapped > 15 {
        return salt - 20_000;
    }

    (((saturated as i64) * 37) + ((wrapped as i64) * 17) + salt) % 1_000_003
}

fn main() {
    let mut total = std::process::id() as i64;

    for i in 0_i32..200_000 {
        total += score(i % 11, i % 6, total % 97);
        total %= 1_000_000_007;
    }

    if total <= 0 {
        std::process::exit(1);
    }
}
